doc/compiler/runtime/Recompilation.md

```diff
@@ -0,0 +1,142 @@
+or the Apache License, Version 2.0 which accompanies this distribution and
+is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+This Source Code may also be made available under the following
+Secondary Licenses when the conditions for such availability set
+forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Background
+
+A natural extension of a JIT Compiler's ability to compile methods "just 
+in time" is to *re*compile methods "just in time". There are several 
+circumstances in when a method might be recompiled; these are outlined 
+in this doc.
+
+# Reasons for recompilation
+
+This section outlines the various reasons for the
+recompilation of a method. Recompilation can either be optional or 
+mandatory. However, since since any compilation is technically not
+mandatory for functional correctness (it exists for performance
+only), both *optional* and *mandatory* in the previous sentence should
+be suffixed with "in order to continue executing the existing compiled
+body".
+
+## Optional Recompilation
+
+The following trigger a recompilation that is optional in the sense that
+it is not functionally incorrect to continue executing the existing JIT'd 
+body but there is a performance benefit to executing a more optimized body. 
+
+### Sampling
+
+The JIT has a Sampler Thread that is used to periodically perform tasks. One
+of these tasks is to set a flag on all Java Threads to yield at the next 
+yield point. Once the threads enter the yield point, they temporarily stop
+executing java code, and instead execute a JIT callback (see 
+`jitMethodSampleInterrupt` in `HookedByTheJit.cpp`). This callback is used
+to build a window of samples, which is used to determine whether or not a 
+method should be recompiled.
+
+### Counting
+
+This mechanism refers to the notion of triggering a recompilation by counting
+the number of times the method is invoked. Guarded Counting Recompilation (GCR)
+is an example of this, wherein the compiler will generate Trees at the start of
+of a method to decrement a counter when the method is invoked, and to trigger
+a recompilation when the counter hits 0. A "Counting Body" is another example
+which is similar in principle to GCR, but has a different pre-prologue. 
+
+### Explict
+
+This section refers to mechanisms that trigger recompilation explicitly, i.e.
+not based on counting or sampling, but via some other criteria. The features
+in OpenJ9 that might explicitly trigger recompilation are:
+* Exception Driven Optimization (EDO)
+* JProfiling
+* On Stack Replacement (OSR)
```

Comment:
Not all OSR will recompile only some - depends on the reason for the OSR. Further only voluntary OSR can do this - involuntary is involuntary so we have no opportunity to request a recompile.

Reply:
Further in the doc, I wrote
>As part of this transition, the mechanism might also trigger a recompilation if this path contributed

so as to suggest that it is possible for some form of OSR to trigger a recompilation; I didn't want to get into the details as the purpose of this doc is to focus on the recompilation aspect, rather than the details of the triggers.